client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2017 AsyncHttpClient Project. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at
+ *     http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package org.asynchttpclient.cookie;
+
+import io.netty.handler.codec.http.cookie.Cookie;
+import org.asynchttpclient.uri.Uri;
+
+import java.util.*;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+public class ThreadSafeCookieStore implements CookieStore {
+
+    private Map<CookieKey, StoredCookie> cookieJar = new HashMap<>();
+    private ReentrantLock lock = new ReentrantLock();
```

Comment:
What are the pros of this `HashMap` + `ReentrantLock`? Why not a `ConcurrentHashMap`?

Reply:
I actually used ReadWriteLock first but then thought it's overkill so I replaced it with simpler class. I can use ConcurrentHashMap if you prefer that.